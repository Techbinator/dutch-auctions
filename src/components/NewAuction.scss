@import "../config/scss/variables.scss";
@import "./AuctionListItem.scss";
.new-auction {
  @extend .auction-list-item;
  .icon-container {
    min-height: 2.58em;
    width: 100%;
    text-align: center;
    font-size: 4em;
    padding-top: 0.5em;
    color: $secondary;
  }
}
