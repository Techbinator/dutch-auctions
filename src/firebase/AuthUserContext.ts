import * as React from "react";
import { User } from "firebase";
const defaultData: User | null = null;
export const AuthUserContext = React.createContext(defaultData);
