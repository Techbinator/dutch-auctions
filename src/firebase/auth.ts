import { auth } from "./config";

// Sign Up
export const createUserWithEmailAndPassword = (
  email: string,
  password: string
) => auth.createUserWithEmailAndPassword(email, password);

// Sign In
export const signInWithEmailAndPassword = (email: string, password: string) =>
  auth.signInWithEmailAndPassword(email, password);

// Sign out
export const signOut = () => auth.signOut();
