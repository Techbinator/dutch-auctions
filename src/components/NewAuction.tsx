import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faPlus } from "@fortawesome/free-solid-svg-icons";
import { User } from "firebase";
import { db } from "../firebase";
import Button from "./Button";

import "./NewAuction.scss";

interface NewAuctionProps {
  authUser: User;
}
interface INewAuctionState {
  title: string;
  startingBid: number;
  error?: string | null;
  loading: boolean;
}

const INITIAL_STATE = {
  title: "",
  startingBid: 1,
  error: null,
  loading: false
};

export default class NewAuction extends React.Component<
  NewAuctionProps,
  INewAuctionState
> {
  state = INITIAL_STATE;
  onSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    this.setState({ loading: true });

    const { title, startingBid } = this.state;
    const { authUser } = this.props;

    db.createNewAuction({ title, startingBid, ownerId: authUser.uid })
      .then(() => {
        this.setState({ ...INITIAL_STATE });
      })
      .catch((error: Error) => {
        this.setState({ error: error.message, loading: false });
      });
  };
  onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    //make typescript happy
    switch (event.target.name) {
      case "title":
        this.setState({ title: event.target.value });
        break;
      case "startingBid":
        this.setState({ startingBid: parseFloat(event.target.value) });
        break;
    }
  };

  render() {
    const { title, startingBid, error, loading } = this.state;
    return (
      <div className="new-auction box-shadow-hover">
        <div className="icon-container">
          <FontAwesomeIcon icon={faPlus} />
        </div>
        <form onSubmit={this.onSubmit}>
          {error && <p className="error">{error}</p>}
          <input
            type="string"
            placeholder="Title"
            name="title"
            value={title}
            onChange={this.onChange}
            required
            minLength={1}
            maxLength={50}
          />
          <input
            name="startingBid"
            value={startingBid}
            type="number"
            onChange={this.onChange}
            placeholder="Price"
            min="1"
            max="10000000"
            step="0.01"
          />
          <Button
            type="primary"
            text={loading ? "Loading..." : "New Auction"}
          />
        </form>
      </div>
    );
  }
}
