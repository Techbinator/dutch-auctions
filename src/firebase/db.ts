import { db } from "./config";
import { User, firestore } from "firebase";

interface IGetAuctions {
  authUser: User;
  currentUserAuctions: boolean;
}
export const getAuctions = ({
  authUser,
  currentUserAuctions
}: IGetAuctions) => {
  let auctions = db.collection("auctions").orderBy("endDate", "desc");
  console.log("here");
  if (authUser && currentUserAuctions) {
    auctions = auctions.where(
      "ownerId",
      "==",
      authUser.uid
    ) as firestore.CollectionReference;
  }

  return auctions;
};

interface ICreateNewAuction {
  title: string;
  startingBid: number;
  ownerId: string;
}

export const createNewAuction = (data: ICreateNewAuction) => {
  var startDate = new Date();
  var endDate = new Date();
  endDate.setMinutes(endDate.getMinutes() + 5);
  const auction = {
    title: data.title,
    startDate: startDate.getTime(),
    endDate: endDate.getTime(),
    startingBid: data.startingBid,
    ownerId: data.ownerId,
    currentMaxBid: 0,
    maxBidder: {
      email: "",
      Id: ""
    }
  };
  return db.collection("auctions").add(auction);
};
