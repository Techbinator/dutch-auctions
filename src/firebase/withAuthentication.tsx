import * as React from "react";
import { firebase } from "./";
import { AuthUserContext } from "./AuthUserContext";
import { User } from "firebase";

interface InterfaceState {
  authUser?: User | null;
}

export const withAuthentication = (Component: any) => {
  class WithAuthentication extends React.Component<{}, InterfaceState> {
    state = {
      authUser: null
    };

    public componentDidMount() {
      firebase.auth.onAuthStateChanged(authUser => {
        authUser
          ? this.setState(() => ({ authUser }))
          : this.setState(() => ({ authUser: null }));
      });
    }

    public render() {
      const { authUser } = this.state;

      return (
        <AuthUserContext.Provider value={authUser}>
          <Component authUser={authUser} />
        </AuthUserContext.Provider>
      );
    }
  }
  return WithAuthentication;
};
