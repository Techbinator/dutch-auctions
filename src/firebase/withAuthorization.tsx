import * as React from "react";
import { withRouter } from "react-router-dom";
import { firebase } from "./";
import Loading from "../components/Loading";
import { AuthUserContext } from "./AuthUserContext";

//verify if user is loged in or not
export const withAuthorization = (checkIsLogedin: boolean) => (
  Component: any
) => {
  class WithAuthorization extends React.Component<any, {}> {
    state = {
      loading: true
    };
    _isMounted: boolean = false;
    public componentDidMount() {
      this._isMounted = true;
      const { history } = this.props;
      firebase.auth.onAuthStateChanged(authUser => {
        if (checkIsLogedin && !authUser) {
          history.push("/login");
        } else if (!checkIsLogedin && authUser) {
          history.push("/my_auctions");
        }
        if (this._isMounted) {
          this.setState({ loading: false });
        }
      });
    }

    public componentWillUnmount() {
      this._isMounted = false;
    }

    public render() {
      const { loading } = this.state;
      if (loading) {
        return <Loading />;
      }

      return (
        <AuthUserContext.Consumer>
          {authUser => <Component authUser={authUser} {...this.props} />}
        </AuthUserContext.Consumer>
      );
    }
  }

  return withRouter(WithAuthorization);
};
