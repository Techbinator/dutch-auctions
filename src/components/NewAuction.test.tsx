import React from "react";
import { shallow } from "enzyme";
import NewAuction from "./NewAuction";
import authUser from "../helpers/__mocks__/authUser.json";
import { User } from "firebase";
import { db } from "../firebase";

describe("Loading component", () => {
  let wrapper: any;
  beforeEach(() => {
    wrapper = shallow(<NewAuction authUser={authUser as User} />);
  });

  it("should render snapshot with default values", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should modify state depending on the values selected", () => {
    const titleInput = wrapper.find('input[name="title"]');
    const startingBidInput = wrapper.find('input[name="startingBid"]');

    titleInput.simulate("change", {
      target: { value: "Title value", name: "title" }
    });
    expect(wrapper.state("title")).toBe("Title value");

    startingBidInput.simulate("change", {
      target: { value: "string", name: "startingBid" }
    });
    expect(wrapper.state("startingBid")).toBe(NaN);

    startingBidInput.simulate("change", {
      target: { value: 22, name: "startingBid" }
    });
    expect(wrapper.state("startingBid")).toBe(22);
  });

  it("should submit correct values to firestore", () => {
    const titleInput = wrapper.find('input[name="title"]');
    const startingBidInput = wrapper.find('input[name="startingBid"]');
    const addMock = jest.spyOn(db, "createNewAuction");
    const event = {
      preventDefault: jest.fn()
    } as any;

    titleInput.simulate("change", {
      target: { value: "Title value", name: "title" }
    });

    startingBidInput.simulate("change", {
      target: { value: 22, name: "startingBid" }
    });

    wrapper.instance().onSubmit(event);
    expect(addMock).toBeCalledWith({
      ownerEmail: "dummyemail",
      ownerId: "dummyuuid",
      startingBid: 22,
      title: "Title value"
    });
  });
});
