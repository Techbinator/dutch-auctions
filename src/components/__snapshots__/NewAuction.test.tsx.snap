// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading component should render snapshot with default values 1`] = `
<div
  className="new-auction box-shadow-hover"
>
  <div
    className="icon-container"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            448,
            512,
            Array [],
            "f067",
            "M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z",
          ],
          "iconName": "plus",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
  </div>
  <form
    onSubmit={[Function]}
  >
    <input
      maxLength={50}
      minLength={1}
      name="title"
      onChange={[Function]}
      placeholder="Title"
      required={true}
      type="string"
      value=""
    />
    <input
      max="10000000"
      min="5"
      name="startingBid"
      onChange={[Function]}
      placeholder="Price"
      step="1"
      type="number"
      value={5}
    />
    <_default
      text="New Auction"
      type="primary"
    />
  </form>
</div>
`;
